### ISSUE TITLE
New component: vcenterreceiver

### ISSUE THREAD

**Issue by schmikei** [open-telemetry/opentelemetry-collector-contrib#8377]
## The purpose and use-cases of the new component

I'd like to contribute a vcenter server receiver that monitors the different levels of resources and logs. This receiver should be capable of monitoring ESXi hosts, virtual machines, cluster, resource pool, and datacenter levels of metrics. Along with this receiver it should be able to support both SAN and vSAN storage metrics. 

Along with this, I'd like to use the [opentelemetry-log-collection ](https://github.com/open-telemetry/opentelemetry-log-collection) to help process raw tcp input and then parse them as syslog messages after filtering inproperly formatted entries forwarded from the vCenter server and ESXi hosts.

#### Collection Methods
- Metrics will be sourced via usage of the [govmomi](https://github.com/vmware/govmomi) library to maintain connections and perform queries.
- Logs will be ingested via usage of the [opentelemetry-log-collection ](https://github.com/open-telemetry/opentelemetry-log-collection) library 

#### Supported Versions of vCenter (derived from govmomi)
- 6.5
- 6.7
- 7.0

## Example configuration for the component

```yaml
  vcenter:
    endpoint: <scheme(http|https)>://vcsa.somehost.localnet # address of the vCenter server or SDK enabled ESXi host
    username: username@somehost.localnet # username of a minimally provided Read only user
    password: password # password of the user
    collection_interval: 1m

    syslog_listen_addr: 0.0.0.0:5140 # port to listen on for syslog messages
    max_log_size: 1MB # max log size to be ingested over syslog
    encoding: utf-8 # encoding
    attributes: {} # attributes to attach to entries
    resource: {} # resource info to attach to entries
    tls: # configtls settings
      insecure: false
      insecure_skip_verify: true
```

## Telemetry data types supported

- metrics
- logs

## Sponsor (Optional)

- Needs a sponsor

**Comment by djaglowski** [comment]
Is there anything special about the syslogs e.g. additional content that can be parsed because we know it is coming from vmware? If not, I don't think this receiver needs to support logs. (syslogreceiver can be used directly)

I'm curious to hear more about that, but if this component were to support logs, I think that instead of using opentelemetry-log-collection, it would be more appropriate to wrap `syslogreceiver`. (See conversation [over here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8370) about wrapping components.) Ultimately, this would use the log collection library, but I don't think we should have any more components than necessary interacting with it directly.

In any case, it's not clear to me how the proposed config would support syslog anyways. Is there a missing endpoint? This maybe brings up interesting questions about whether metrics and logs parameters should be mixed together or split into sub-sections of the config.

**Comment by schmikei** [comment]
> Is there anything special about the syslogs e.g. additional content that can be parsed because we know it is coming from vmware? If not, I don't think this receiver needs to support logs. (syslogreceiver can be used directly)
> 
> I'm curious to hear more about that, but if this component were to support logs, I think that instead of using opentelemetry-log-collection, it would be more appropriate to wrap `syslogreceiver`. (See conversation [over here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8370) about wrapping components.) Ultimately, this would use the log collection library, but I don't think we should have any more components than necessary interacting with it directly.
> 


As far as what we've seen in the past for vCenter forwarded logs, not all messages  generally inputted as raw tcp input i.e 
```txt
23 <134>1 2022-03-01T19:34:33.160936+00:00 vcsa vapi-endpoint-access - - - 2022-03-01T19:34:33.160Z | jetty-default-29          | 127.0.0.1 - - [01/Mar/2022:19:34:29 +0000] "GET /rest/com/vmware/vapi/metadata/metamodel/component?~method=OPTIONS HTTP/1.1" 200 2534 "-" "python-requests/2.19.1" 3558
```

note the leading extra ID (23) prefixed with it which breaks the `rfc5424` format. I can look more into when these show up to see if it is something else configurable with vCenter logs. Definitely something that may need some 


As far as what actually will be used, yes I think a more accurate proposal of what I want to do is to wrap the [tcplogreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/tcplogreceiver) and then use the [operators](https://github.com/open-telemetry/opentelemetry-log-collection/blob/main/docs/operators/README.md#what-operators-are-available) to do the syslog parsing once input has been partially sanitized. @djaglowski  does that seem like the approach you would take here as well?

 
> In any case, it's not clear to me how the proposed config would support syslog anyways. Is there a missing endpoint? This maybe brings up interesting questions about whether metrics and logs parameters should be mixed together or split into sub-sections of the config.

For this one I am also somewhat curious, as far as I can tell there haven't been too many components that do this shared config across logging and metrics and it makes me ponder config conflicts such as `tls`.

```yaml
receivers:
  vcenter:
    ### Metrics Config ###
    endpoint: <scheme(http|https)>://vcsa.somehost.localnet
    username: username@somehost.localnet
    password: password 
    collection_interval: 1m
    ### Logging Config ###
    syslog_listen_addr: 0.0.0.0:5140
    max_log_size: 1MB
    encoding: utf-8
    attributes: {}
    resource: {}
    ### Shared(? or perhaps a logging_tls) ###
    tls:
      insecure: false
      insecure_skip_verify: true

exporters:
  logging:
    loglevel: debug

pipelines:
  metrics:
    receivers: [vcenter]
    exporters: [logging]
  logging:
    receiver: [vcenter]
    exporters: [logging]
```

I can imagine something like this but I can imagine it gets a tad hairy when config for the receiver is combined. Do you have any thoughts on what is the correct way of solving this issue where its not always just raw syslog and may need some custom processing?

**Comment by djaglowski** [comment]
> As far as what actually will be used, yes I think a more accurate proposal of what I want to do is to wrap the [tcplogreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/tcplogreceiver) and then use the [operators](https://github.com/open-telemetry/opentelemetry-log-collection/blob/main/docs/operators/README.md#what-operators-are-available) to do the syslog parsing once input has been partially sanitized. @djaglowski does that seem like the approach you would take here as well?

I see. You want to receive the logs and, possibly clean them up a bit, and then feed them through the syslog parser. Roughly the same consideration would apply. I think you should use the `tcplog` receiver, rather than using the log-collection library's `tcp_input` directly.

Does vcenter necessarily send logs via tcp though? Could it be udp?

I think the mixed config is messy and should be split up. Roughly:
```yaml
vcenter:
  metrics:
    endpoint: <scheme(http|https)>://vcsa.somehost.localnet
    username: username@somehost.localnet
    password: password 
    collection_interval: 1m
    tls:
      insecure: false
      insecure_skip_verify: true
  logs:
    listen_addr: 0.0.0.0:5140
    max_log_size: 1MB
    encoding: utf-8
    tls:
      insecure: false
      insecure_skip_verify: true
```

You can then easily determine whether or not a metrics or logs receiver should be created.

One issue with this is that it collides with a well established pattern in scrapers, where `metrics` refers to an [embedded](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/1b4d33f6eb5b3d774a946866405cc4e11025b5cd/receiver/apachereceiver/config.go#L31) `metadata.MetricSettings` struct, which allows the user to turn on/off individual metrics. Probably we need to target something like:

```yaml
vcenter:
  metrics:
    endpoint: <scheme(http|https)>://vcsa.somehost.localnet
    username: username@somehost.localnet
    password: password 
    collection_interval: 1m
    tls:
      insecure: false
      insecure_skip_verify: true
    settings:
      ...
  logs:
    listen_addr: 0.0.0.0:5140
    max_log_size: 1MB
    encoding: utf-8
    tls:
      insecure: false
      insecure_skip_verify: true
```

**Comment by djaglowski** [comment]
How will integration testing work? It looks like `govmomi` provides a [simulated API](https://github.com/vmware/govmomi/tree/master/vcsim). Can we run that in a container and run meaningful tests against it?

**Comment by djaglowski** [comment]
> works great for doing at least property collection

I think you are talking about static configuration values such as total memory, right? What about dynamic metric values like used memory?

**Comment by schmikei** [comment]
Properties are not necessarily always static they get periodically updated. Here is a link for example the [Virtual Machine Documentation](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/9fd87c06-14a3-41e5-b28d-277864a80f29/d6112c2a-b124-4fa5-96d7-9fb4b6f1bb50/doc/vim.vm.Summary.QuickStats.html)  that you retrieve via the Property Collector which in the cases I am thinking of are updated in near real time of when you retrieve the property. 

However that is only one source of potential metrics as there is the [Performance Manager API](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/1ef6c336-7bef-477d-b9bb-caa1767d7e30/82521f49-9d9a-42b7-b19b-9e6cd9b30db1/vim.PerformanceManager.html). As well as other APIs for more real time performance data like the [vSAN performance manager API](https://vdc-download.vmware.com/vmwb-repository/dcr-public/3325c370-b58c-4799-99ff-58ae3baac1bd/45789cc5-aba1-48bc-a320-5e35142b50af/doc/vim.cluster.VsanPerformanceManager.html).

On a side I was able to find the [vSAN equivalent simulator](https://github.com/vmware/govmomi/blob/master/vsan/simulator/simulator.go). And should be able to use that to write more tests.

**Comment by djaglowski** [comment]
> Properties are not necessarily always static they get periodically updated. Here is a link for example the [Virtual Machine Documentation](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/9fd87c06-14a3-41e5-b28d-277864a80f29/d6112c2a-b124-4fa5-96d7-9fb4b6f1bb50/doc/vim.vm.Summary.QuickStats.html) that you retrieve via the Property Collector which in the cases I am thinking of are updated in near real time of when you retrieve the property.

Thanks for clarifying @schmikei.

> However that is only one source of potential metrics as there is the [Performance Manager API](https://vdc-repo.vmware.com/vmwb-repository/dcr-public/1ef6c336-7bef-477d-b9bb-caa1767d7e30/82521f49-9d9a-42b7-b19b-9e6cd9b30db1/vim.PerformanceManager.html). As well as other APIs for more real time performance data like the [vSAN performance manager API](https://vdc-download.vmware.com/vmwb-repository/dcr-public/3325c370-b58c-4799-99ff-58ae3baac1bd/45789cc5-aba1-48bc-a320-5e35142b50af/doc/vim.cluster.VsanPerformanceManager.html).

Do these APIs share an endpoint & authentication?
 
> On a side I was able to find the [vSAN equivalent simulator](https://github.com/vmware/govmomi/blob/master/vsan/simulator/simulator.go). And should be able to use that to write more tests.

That's great. Seems it is sufficiently testable then.

**Comment by schmikei** [comment]
> Do these APIs share an endpoint & authentication?

Most of it is handled via [authenticated user session authentication in govmomi](https://github.com/vmware/govmomi/blob/9054158bb458f8c383422253709b56dc318ba2c0/session/manager.go#L87) which gives a central authenticated client to go off of to make the API calls. 

As far as endpoints are concerned there are multiple `routes` that govmomi will reach out to when making the underlying calls but communication is only going to `one single ESXi host or vCenter` running the appropriate management SDK.

**Comment by djaglowski** [comment]
Ok, sounds like the metric receiver can operator on one endpoint and one set of credentials then.

**Comment by djaglowski** [comment]
This looks like it will be a somewhat complicated receiver, but it seems fully testable. VMware support in the collector would be a great addition. I'm happy to sponsor this.

Will you submit the metrics and logs parts separately?

**Comment by djaglowski** [comment]
Closed by #9224
